File: Tutorial2CRUDOperations/src/main/java/com/example/tutorial2crudoperations/data/source/UserRepository.java
package com.example.tutorial2crudoperations.data.source;

import com.example.tutorial2crudoperations.data.User;

import java.util.List;

public class UserRepository implements UserDataSource {

    private static UserRepository INSTANCE = null;

    private UserDataSource mLocalUserDataSource;

    private UserRepository(UserDataSource userDataSource) {
        mLocalUserDataSource = userDataSource;
    }

    public static UserRepository getInstance(UserDataSource userDataSource) {
        if (INSTANCE == null) {
            INSTANCE = new UserRepository(userDataSource);
        }
        return INSTANCE;
    }

    @Override
    public void insertUser(User user) {
        mLocalUserDataSource.insertUser(user);
    }

    @Override
    public void insertUsers(User... users) {
        mLocalUserDataSource.insertUsers(users);

    }

    @Override
    public void insertBothUsers(User user1, User user2) {
        mLocalUserDataSource.insertUsers(user1, user2);
    }

    @Override
    public void insertUsersAndFriends(User user, List<User> friends) {
        mLocalUserDataSource.insertUsersAndFriends(user, friends);
    }

    @Override
    public void updateUsers(User... users) {
        mLocalUserDataSource.insertUsers(users);
    }

    @Override
    public void deleteUsers(User... users) {
        mLocalUserDataSource.insertUsers(users);
    }

    @Override
    public int countUsers() {
        return mLocalUserDataSource.countUsers();
    }

    @Override
    public User[] loadAllUsers() {
        return mLocalUserDataSource.loadAllUsers();
    }

    @Override
    public User[] loadAllUsersOlderThan(int minAge) {
        return mLocalUserDataSource.loadAllUsersOlderThan(minAge);
    }

    @Override
    public User[] loadAllUsersBetweenAges(int minAge, int maxAge) {
        return mLocalUserDataSource.loadAllUsersBetweenAges(minAge, maxAge);
    }

    @Override
    public List<User> findUserWithName(String search) {
        return mLocalUserDataSource.findUserWithName(search);
    }

    @Override
    public User getUser(long userId) {
        return mLocalUserDataSource.getUser(userId);
    }

    @Override
    public List<User> loadAllByIds(int[] userIds) {
        return mLocalUserDataSource.loadAllByIds(userIds);
    }

    @Override
    public User findByName(String first, String last) {
        return mLocalUserDataSource.findByName(first, last);
    }

    @Override
    public int deleteAll() {
        return mLocalUserDataSource.deleteAll();
    }
}

File: Tutorial2-1CRUDOperations/src/main/java/com/example/tutorial2crudoperations/data/source/UserRepository.kt
package com.example.tutorial2crudoperations.data.source

import com.example.tutorial2crudoperations.data.User

// TODO Do User Data Caching inside this class
class UserRepository private constructor(private val mLocalUserDataSource: UserDataSource) :
    UserDataSource {
    override fun insertUser(user: User) {
        mLocalUserDataSource.insertUser(user)
    }

    override fun insertUsers(vararg users: User) {
        mLocalUserDataSource.insertUsers(*users)
    }

    override fun insertBothUsers(
        user1: User,
        user2: User
    ) {
        mLocalUserDataSource.insertUsers(user1, user2)
    }

    override fun insertUsersAndFriends(
        user: User,
        friends: List<User>
    ) {
        mLocalUserDataSource.insertUsersAndFriends(user, friends)
    }

    override fun updateUsers(vararg users: User) {
        mLocalUserDataSource.insertUsers(*users)
    }

    override fun deleteUsers(vararg users: User) {
        mLocalUserDataSource.insertUsers(*users)
    }

    override fun countUsers(): Int {
        return mLocalUserDataSource.countUsers()
    }

    override fun loadAllUsers(): Array<User> {
        return mLocalUserDataSource.loadAllUsers()
    }

    override fun loadAllUsersOlderThan(minAge: Int): Array<User> {
        return mLocalUserDataSource.loadAllUsersOlderThan(minAge)
    }

    override fun loadAllUsersBetweenAges(
        minAge: Int,
        maxAge: Int
    ): Array<User> {
        return mLocalUserDataSource.loadAllUsersBetweenAges(minAge, maxAge)
    }

    override fun findUserWithName(search: String): List<User> {
        return mLocalUserDataSource.findUserWithName(search)
    }

    override fun getUser(userId: Long): User {
        return mLocalUserDataSource.getUser(userId)
    }

    override fun loadAllByIds(userIds: Array<Int>): List<User> {
        return mLocalUserDataSource.loadAllByIds(userIds)
    }

    override fun findByName(
        first: String,
        last: String
    ): User {
        return mLocalUserDataSource.findByName(first, last)
    }

    override fun deleteAll(): Int {
        return mLocalUserDataSource.deleteAll()
    }

    companion object {
        private lateinit var INSTANCE: UserRepository

        @JvmStatic
        fun getInstance(userDataSource: UserDataSource): UserRepository {

            if (INSTANCE == null) {
                INSTANCE = UserRepository(userDataSource)
            }

            return INSTANCE
        }
    }

}